sdk/guides/agent-server/custom-tools.mdx

```diff
@@ -0,0 +1,510 @@
+
+<Note>
+For standalone custom tools (without remote agent server), see the [Custom Tools guide](/sdk/guides/custom-tools).
+</Note>
+
+<Warning>
+**Binary Mode Limitation**: Custom tools only work with **source mode** deployments of the agent server. If you're using a pre-built binary agent server (built with PyInstaller), custom tools cannot be dynamically imported at runtime because the binary's frozen Python interpreter can only load modules that were bundled at build time.
+
+When using `DockerDevWorkspace`, the agent server is built in source mode by default, so custom tools work correctly. However, if you're deploying a standalone binary agent server, you'll need to either:
+- Use source mode instead (recommended for custom tools)
+- Rebuild the binary with your custom tools included at build time
+
+See [GitHub issue #1531](https://github.com/OpenHands/software-agent-sdk/issues/1531) for more details and alternative approaches.
+</Warning>
+
+## Use Cases
+
+Custom tools with remote agent servers are useful for:
+
+- **Structured data collection**: Define tools like `log_data`, `record_metric`, or `track_event` to collect structured data during agent runs
+- **Custom integrations**: Tools that interact with external systems (APIs, databases, etc.)
+- **Domain-specific operations**: Business logic tools specific to your application
+- **Downstream processing**: Collected data can be used to generate reports, trigger workflows, etc.
+
+## Architecture
+
+```
+┌─────────────────┐         ┌──────────────────────────┐
+│   SDK Client    │         │   Remote Agent Server    │
+│                 │         │   (Built on custom base) │
+│  - Define tools │◄────────┤                          │
+│  - Send tasks   │   API   │  - Custom tools in       │
+│  - Get results  │         │    Python path           │
+│                 │         │  - Dynamic registration  │
+└─────────────────┘         │  - Tool execution        │
+                            │  - JSON file output      │
+                            └──────────────────────────┘
+```
+
+## Complete Example
+
+<Note>
+A complete working example is available on GitHub: [examples/02_remote_agent_server/05_custom_tool/](https://github.com/OpenHands/software-agent-sdk/tree/main/examples/02_remote_agent_server/05_custom_tool)
+</Note>
+
+The example includes:
+- **Custom tool implementation** (`custom_tools/log_data.py`) - A tool for logging structured data to JSON
+- **Dockerfile** - Simple Dockerfile that copies custom tools into a base image
+- **Build script** (`build_custom_image.sh`) - Builds the custom base image
+- **SDK example** (`custom_tool_example.py`) - Demonstrates the full workflow
+
+## Step-by-Step Guide
+
+### 1. Define Your Custom Tool
+
+Create a Python module for your tool (e.g., `custom_tools/log_data.py`):
+
+```python
```

Comment:
@OpenHands please make this PR more succinct - you only need to mention the core part in this doc. Please refer to the corresponding PR to understand the context